### DISCUSSION TITLE
Faultcode and detail in the body of a Fault response please?

### DISCUSSION THREAD

**Discussion by temirovazat** [CoreWCF/CoreWCF#641] [Q&A]
Hi, what is the correct way of customising faultstring, faultcode and detail in the body of a Fault response please?
I tried to do that using ProvideFault in many ways, but had no success.

**Comment by mconnew** [comment]
You do this by throwing `FaultException` if you just want to provide the reason and fault code (pass them to the constructor). If you also want to provide a body, then you use `FaultException<TDetail>`. You create a class to provide a TDetail and it gets serialized with DataContractSerializer to provide the body. You can use the DataContract and DataMember attributes to modify the XML names and namespaces if you need. If you are okay with the defaults, then you don't need to attribute anything and all public properties will be serialized. On your operation you need to attribute it with FaultContract providing the type of TDetail otherwise CoreWCF will just send a generic "Something went wrong" fault message. A good example can be found in the WCF docs [here](https://docs.microsoft.com/en-us/dotnet/framework/wcf/samples/fault-contract), just change the using from System.ServiceModel to CoreWCF.

**Comment by projecteon** [comment]
I am not able to make this work, (not even with the simple [sample code](https://github.com/CoreWCF/samples/tree/main/Scenarios/Getting-Started-with-CoreWCF)). I also tried adding `[ServiceBehavior(IncludeExceptionDetailInFaults = true)]`attribute to the service without it making any difference. I also tried updating the code to he latest nuget packages.
I only seem to receive the basic FaultException.
```csharp
var factory = new ChannelFactory<IEchoService>(binding, new EndpointAddress($"{hostAddr}/EchoService/basicHttp"));
factory.Open();
try
{
    IEchoService client = factory.CreateChannel();
    channel = client as IClientChannel;
    channel.Open();
    var result = await client.FailEcho("Hello World!");
    channel.Close();
    Console.WriteLine(result);
}
catch (FaultException<EchoFault> ex)
{
    // I should end up here
    Console.WriteLine($"Failed with '{ex.Message}'");
}
catch (FaultException ex)
{
    // I always end up here....
    Console.WriteLine($"Failed with '{ex.Message}'");
}
catch (Exception ex)
{
    Console.WriteLine($"Failed with '{ex.Message}'");
}
finally
{
    factory.Close();
}
```

**Reply by mconnew** [reply]
You need to attribute your operations with FaultContractAttribute to make everything work. Did you do that?

**Reply by projecteon** [reply]
Yes I did
As I said, this is using the sample code for CoreWCF ([link](https://github.com/CoreWCF/samples/blob/main/Scenarios/Getting-Started-with-CoreWCF/NetCoreClient/IEchoService.cs)) and it has the attribute. I just changed the code on the client side to add a call to the fault triggering endpoint. I only end up getting the generic FaultException.

**Comment by blorger** [comment]
I had similar problems when I used FaultException from System.ServiceModel instead from CoreWCF.
Although, service would still return fault details when I decorated service implementation with [ServiceBehavior(IncludeExceptionDetailInFaults = true)].  Of course attribute was from CoreWCF not System.ServiceModel.

**Reply by projecteon** [reply]
Sample code for CoreWCF does use the System.ServiceModel namespace on the client side.
Is it then not possible to get details when using a net Framework frontend?

**Reply by mconnew** [reply]
It's not clear to me exactly what you're asking. I'll give a brief overview and hopefully that will answer your question.  
For a FaultException\<TDetail\> to be thrown in a service, and then rethrown on the client, the contract interface needs the FaultContract attribute applied specifying the detail type, on the client and on the server. Without this, you will only get the non-generic exception FaultException thrown. We have compatibility code so you can use a System.ServiceModel.FaultContractAttribute on the contract interface on the client side if you need to have a common implementation. You just need a reference to it on the server side (usually the System.ServiceModel.Primitives client nuget package, but if targeting .NET Framework, technically you should be able to get away with referencing System.ServiceModel.dll instead).  
When it comes to implementation of your service, you have to throw a FaultException\<TDetail\> from the CoreWCF namespace. There's code in CoreWCF which handles this specific exception type. I did look into adding compatibility code for this, but I believe I bumped into a problem that I can't remember right now. I think the issue was that you can derive FaultException (eg declare `public class MyException : FaultException<MyDetail>` so you can just throw MyException) which we can't convert over, and there's extensibility points where we pass the FaultException to which then couldn't be done. Anyway, a FaultException\<TDetail\> gets sent to the client as a Fault message (this is a 1st class concept at the SOAP level and isn't just a WCF concept). On the client side, when a Fault message is received, it looks at the details and see if there's a matching FaultContract attribute with the correct detail type, and if there is, creates an instance of System.ServiceModel.FaultException\<TDetail\> from the Fault message and throws it for your client code to catch. As you can see, a FaultException is just a representation of a special type of SOAP message sent over the wire, and the actual types don't need to match between the client and server implementation. The service can throw CoreWCF.FaultException\<Foo\> and the client channel will throw System.ServiceModel.FaultException\<Foo\> (presuming you have your contract attributes applied correctly).  
So to summarize, you can get details sent, you just need to throw/catch either the WCF or CoreWCF namespaced types depending on if your code is service or client code.

**Reply by projecteon** [reply]
Its simple, core WCF has a sample code here: https://github.com/CoreWCF/samples/tree/main/Scenarios/Getting-Started-with-CoreWCF
It sets up the `FaultException<MyDetail>` service call on the server side but never calls it on the client side.
I added the code in my previous post to make a call to the endpoint using the detailed fault exception. It results in just getting the non-generic version (FaultException). I've tried all the of the above as far as I can see, but I still only recieve the non-generic FaultException.

(What Im trying is using a CoreWCF service from a .net Framework(4.8) application, but I cannot seem to create code that works even when both are net8 apps).

**Reply by g7ed6e** [reply]
We actually have tests for this please check https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Http/tests/FaultContractTests.cs.

**Reply by projecteon** [reply]
Ive seen the test, question then remains why it does not seem to work in practice on the official example code.
I went through a lot of the repository code to see if I did something strange or wrong before posting here.

**Comment by fbenoit-coexya** [comment]
I've had the same problem of CoreWCF seemingly ignoring my FaultContract, returning a regular "please use IncludeExceptionDetailsInFaults" fault instead of my FaultException<TDetail>.

Turned out my CoreWCF server's user code was throwing a System.ServiceModel.FaultException<TDetail> instead of a CoreWCF.FaultException<TDetail>. Modifying my code to throw CoreWCF's FaultException fixed the problem.

**Edit:** However, it opened another problem: Unlike System.ServiceModel.Primitives (.NET Standard 2.0 version), CoreWCF.Primitives is very dependency-heavy.
Does CoreWCF support a way to configure, at the WebHost level, something that would convert the System.ServiceModel.FaultExceptions into CoreWCF.FaultException ?

**Reply by fbenoit-coexya** [reply]
Turns out the only reason I had to worry about CoreWCF.FaultException<>'s location was because I was doing something wrong (my WCF service was inheriting a class defined in another assembly and directly exposing its methods instead of wrapping).
But I'm still curious about the possibility of a WebHost-level error filter...

**Reply by mconnew** [reply]
Can you tell me a little more about what you're thinking with a WebHost-level error filter? Maybe some mocked up code of how you're thinking things would work?

**Reply by fbenoit-coexya** [reply]
Sorry for the slow response, I have trouble organizing my thoughts, but I'll try to explain what I had in mind:

- For starters, **perhaps it already exists** somewhere in the ASP.NET Core stack. But does it exist at the *right* step of the pipeline? In the case that interests us, it would need to be somewhere between "my user code throws an exception" and "CoreWCF code turns this exception into a SOAP Fault".
- I was figuring something that would either override parts of the "turn the exception into a SOAP fault" behavior, or allow replacing the exception with another one (or allow other stuff, like logging the exception how I want, etc.).
- Usage-wise, I suppose it would be accessible though something like an `AddErrorFilter()` method in the `IServiceBuilder` interface...

Sorry I don't have anything more precise. 
(For this specific problem it's moot anyway, since I fixed all my service to properly throw a CoreWCF.FaultException<MyDetailsType> and my client code now properly receives the details.)

**Reply by mconnew** [reply]
It sounds like what you want to do is implement an IErrorHandler. [Here's the documentation for WCF](https://learn.microsoft.com/en-us/dotnet/framework/wcf/samples/extending-control-over-error-handling-and-reporting), the api shapes should all be the same, just a different namespace.
